From a562110b1cebbad3ea3196b201004e00c73e87be Mon Sep 17 00:00:00 2001
From: wwidjaja <w.widjaja@intel.com>
Date: Mon, 6 Jan 2020 18:15:40 +0800
Subject: [PATCH] PONRTSYS-5288- PRX300 static NGI config driver

  set NGI permision for SSX1 & SSX4
  specific permissions fix :
  - TA OTP to disable (prevent hang, issue bus error)
  - TA CGU R&W to MIPS 4kec only
  - TA RCU R&W to MIPS 4kec only

PONRTSYS-5288 - cleanup

  - add SPDX license for GPLv2
  - remove if 0 macro block
  - ngi ssx1 & ssx4 simplify

PONRTSYS-5288 - clean up debug print, add return value

PONRTSSYS-5288 - change compat string to vendor,soc-ip

  compat string into "intel,prx300-ngi"
---
 arch/mips/lantiq/grx500_bootcore/Makefile          |   1 +
 arch/mips/lantiq/grx500_bootcore/ngi_prx300.c      | 331 +++++++++++++++++++++
 arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx1.h |  98 ++++++
 arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx4.h |  53 ++++
 4 files changed, 483 insertions(+)

diff --git a/arch/mips/lantiq/grx500_bootcore/Makefile b/arch/mips/lantiq/grx500_bootcore/Makefile
index 99a1323d8cd0..662ccde23070 100644
--- a/arch/mips/lantiq/grx500_bootcore/Makefile
+++ b/arch/mips/lantiq/grx500_bootcore/Makefile
@@ -1,5 +1,6 @@
 obj-y := prom.o clk.o sysctrl.o early_printk.o irq.o
 obj-$(CONFIG_SOC_GRX500_BOOTCORE) += ngi.o
+obj-$(CONFIG_SOC_PRX300_BOOTCORE) += ngi_prx300.o
 #obj-y := prom.o clk.o irq.o sysctrl.o early_printk.o ngi.o timer.o
 
 ifdef CONFIG_RESET_IAP
diff --git a/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c b/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c
new file mode 100644
index 000000000000..16cac8d6855e
--- /dev/null
+++ b/arch/mips/lantiq/grx500_bootcore/ngi_prx300.c
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ * William Widjaja <w.widjaja@intel.com>
+ */
+
+#include <asm/io.h>
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/of_platform.h>
+
+/* ssx register defintion */
+#include "ngi_prx300_ssx1.h"
+#include "ngi_prx300_ssx4.h"
+
+#define NONE		0x0
+#define M4KEC_ONLY	0x1
+#define CPU0		0x2
+#define CPU1		0x4
+#define CPU2		0x8
+#define CPU3		0x10
+#define OTHERS		0x20
+#define	EIP123		0x40
+#define IAP_N_4KEC	M4KEC_ONLY | CPU0 | CPU1 | CPU2 | CPU3
+#define	GRP0_TO_4	M4KEC_ONLY | CPU0 | CPU1 | CPU2 | CPU3 | OTHERS
+#define	ALL_CPUS	GRP0_TO_4 | 0x40
+
+#define SSX_REGSIZE 0x00100000
+
+static u32 permission_wr(void __iomem* base, u32 offset, u32 val)
+{
+	__raw_writel(val, (volatile void __iomem *)(base+offset));
+	return 0;
+}
+
+void set_permission(void)
+{
+	void __iomem* ssx1_base = 0;
+	void __iomem* ssx4_base = 0;
+
+	ssx1_base = ioremap_nocache( SSX1_MODULE_BASE,             SSX_REGSIZE);
+	if (NULL == ssx1_base) {
+		pr_err("prx300-ngi: Error ioremap sxx1!");
+		return;
+	}
+
+	ssx4_base = ioremap_nocache( SSX4_SHARED_LINK_MODULE_BASE, SSX_REGSIZE);
+	if (NULL == ssx4_base) {
+		pr_err("prx300-ngi: Error ioremap sxx4!");
+		return;
+	}
+
+	/* SSX1 Registers */
+	permission_wr(ssx1_base,
+			REG0_TREG0_PM_TREG0_PM_READ_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx1_base,
+			REG0_TREG0_PM_TREG0_PM_WRITE_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx1_base,
+			REG0_TOTP_PM_TOTP_PM_READ_PERMISSION_0, NONE);
+	permission_wr(ssx1_base,
+			REG0_TOTP_PM_TOTP_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TCBM1_PM_TCBM1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TCBM1_PM_TCBM1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TCBM2_PM_TCBM2_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TCBM2_PM_TCBM2_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TE123_PM_TE123_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TE123_PM_TE123_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMA3_PM_TDM3_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMA3_PM_TDM3_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAT1_PM_TDMAT1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAT1_PM_TDMAT1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TEX04_PM_TEX04_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TEX04_PM_TEX04_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_0, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_0, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_1, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_1, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_2, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_2, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_3, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_3, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_4, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_4, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_5, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_5, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_6, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_6, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_7, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_7, ALL_CPUS);
+	permission_wr(ssx1_base,
+			REG0_TPCTL_PM_TPCTL_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCTL_PM_TPCTL_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPUB_PM_TPUB_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPUB_PM_TPUB_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPON_PM_TPON_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPON_PM_TPON_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TGSWIP_PM_TGSWIP_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TGSWIP_PM_TGSWIP_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TQSPIC_PM_TQSPIC_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TQSPIC_PM_TQSPIC_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TMSI_PM_TMSI1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TMSI_PM_TMSI1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAT2_PM_TDMAT2_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAT2_PM_TDMAT2_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TQSPID_PM_TQSPID_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TQSPID_PM_TQSPID_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAR1_PM_TDMAR1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAR1_PM_TDMAR1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAR2_PM_TDMAR2_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDMAR2_PM_TDMAR2_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TACA_PM_TACA_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TACA_PM_TACA_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCIE_CR1_PM_TPCIE_CR1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCIE_CR1_PM_TPCIE_CR1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCI2_PM_TPCI2_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCI2_PM_TPCI2_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCI1_PM_TPCI1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCI1_PM_TPCI1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TMSI2_PM_TMSI2_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TMSI2_PM_TMSI2_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCIE_CR2_PM_TPCIE_CR2_PM_READ_PERMISSION_0,  GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPCIE_CR2_PM_TPCIE_CR2_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TLNPPV4_PM_TLNPPV4_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TLNPPV4_PM_TLNPPV4_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TXPCS_PM_TXPCS_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TXPCS_PM_TXPCS_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPC1DBI_PM_TPCI1DBI_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPC1DBI_PM_TPCI1DBI_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPC2DBI_PM_TPCI2DBI_PM_READ_PERMISSION_0,  GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TPC2DBI_PM_TPCI2DBI_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TSSB_PM_TSSB_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TSSB_PM_TSSB_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TROM_PM_TROM_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TROM_PM_TROM_PM_WRITE_PERMISSION_0,  GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TMACS_PM_TMAC_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TMACS_PM_TMAC_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TIOCU1_PM_TIOCU1_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TIOCU1_PM_TIOCU1_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDM4_PM_TDM4_PM_READ_PERMISSION_0, GRP0_TO_4);
+	permission_wr(ssx1_base,
+			REG0_TDM4_PM_TDM4_PM_WRITE_PERMISSION_0, GRP0_TO_4);
+
+
+	/* SSX 4 registers */
+	permission_wr(ssx4_base,
+			REG4_TREG4_PM_TREG4_PM_WRITE_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx4_base,
+			REG4_TREG4_PM_TREG4_PM_READ_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx4_base,
+			REG4_TCGU_PM_TCGU_PM_WRITE_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx4_base,
+			REG4_TCGU_PM_TCGU_PM_READ_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx4_base,
+			REG4_TRCU_PM_TRCU_PM_WRITE_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx4_base,
+			REG4_TRCU_PM_TRCU_PM_READ_PERMISSION_0, M4KEC_ONLY);
+	permission_wr(ssx4_base,
+			REG4_TPMU_PM_TPMU_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TPMU_PM_TPMU_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TPCM_PM_TPCM_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TPCM_PM_TPCM_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPIO_PM_TGPIO_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPIO_PM_TGPIO_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TLEDC_PM_TLEDC_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TLEDC_PM_TLEDC_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TI2C_PM_TI2C0_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TI2C_PM_TI2C0_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TASC0_PM_TASC0_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TASC0_PM_TASC0_PM_READ_PERMISSION_0,  IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TASC1_PM_TASC1_PM_WRITE_PERMISSION_0,  IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TASC1_PM_TASC1_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPT0_PM_TGPT0_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPT0_PM_TGPT0_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TSSC0_PM_TSSC0_PM_WRITE_PERMISSION_0,  IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TSSC0_PM_TSSC0_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TSSC1_PM_TSSC1_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TSSC1_PM_TSSC1_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TDMA0_PM_TDM0_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TDMA0_PM_TDM0_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPT1_PM_TGPT1_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPT1_PM_TGPT1_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPT2_PM_TGPT2_PM_WRITE_PERMISSION_0,  IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TGPT2_PM_TGPT2_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TI2C0_PM_TI2C1_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TI2C0_PM_TI2C1_PM_READ_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TI2C1_PM_TI2C2_PM_WRITE_PERMISSION_0, IAP_N_4KEC);
+	permission_wr(ssx4_base,
+			REG4_TI2C1_PM_TI2C2_PM_READ_PERMISSION_0, IAP_N_4KEC);
+
+	return;
+}
+
+static int prx300_ngi_probe(struct platform_device *pdev)
+{
+	set_permission();
+
+	return 0;
+}
+
+static const struct of_device_id prx300_ngi_match[] = {
+	{ .compatible = "intel,prx300-ngi" },
+	{},
+};
+MODULE_DEVICE_TABLE(of, prx300_ngi_match);
+
+ 
+static struct platform_driver prx300_ngi_driver = {
+	.probe = prx300_ngi_probe,
+	.driver = {
+		.name = "prx300-ngi",
+		.owner = THIS_MODULE,
+		.of_match_table = prx300_ngi_match,
+	},
+};
+ 
+int __init prx300_ngi_init(void)
+{
+	int ret;
+ 
+	ret = platform_driver_register(&prx300_ngi_driver);
+	if (ret)
+		pr_info("prx300-ngi: Error registering platform driver!");
+	return ret;
+}
+ 
+core_initcall(prx300_ngi_init);
+
diff --git a/arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx1.h b/arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx1.h
new file mode 100644
index 000000000000..4b5d25b35c4e
--- /dev/null
+++ b/arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx1.h
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ */
+
+#ifndef _SSX1_SSX_H
+#define _SSX1_SSX_H
+
+/*! Base Address of SSX1 */
+#define SSX1_MODULE_BASE 0x18000000u
+
+/*! Permission Offset (relative) */
+#define REG0_TREG0_PM_TREG0_PM_READ_PERMISSION_0		0x80050
+#define REG0_TREG0_PM_TREG0_PM_WRITE_PERMISSION_0		0x80058
+#define REG0_TREG0_PM_TREG0_PM_READ_PERMISSION_1		0x80070
+#define REG0_TREG0_PM_TREG0_PM_WRITE_PERMISSION_1		0x80078
+#define REG0_TOTP_PM_TOTP_PM_READ_PERMISSION_0			0x80C50
+#define REG0_TOTP_PM_TOTP_PM_WRITE_PERMISSION_0			0x80C58
+#define REG0_TCBM1_PM_TCBM1_PM_READ_PERMISSION_0		0x81050
+#define REG0_TCBM1_PM_TCBM1_PM_WRITE_PERMISSION_0		0x81058
+#define REG0_TCBM2_PM_TCBM2_PM_READ_PERMISSION_0		0x81450
+#define REG0_TCBM2_PM_TCBM2_PM_WRITE_PERMISSION_0		0x81458
+#define REG0_TE123_PM_TE123_PM_READ_PERMISSION_0		0x81C50
+#define REG0_TE123_PM_TE123_PM_WRITE_PERMISSION_0		0x81C58
+#define REG0_TDMA3_PM_TDM3_PM_READ_PERMISSION_0			0x82050
+#define REG0_TDMA3_PM_TDM3_PM_WRITE_PERMISSION_0		0x82058
+#define REG0_TDMAT1_PM_TDMAT1_PM_READ_PERMISSION_0		0x82450
+#define REG0_TDMAT1_PM_TDMAT1_PM_WRITE_PERMISSION_0		0x82458
+#define REG0_TEX04_PM_TEX04_PM_READ_PERMISSION_0		0x84050
+#define REG0_TEX04_PM_TEX04_PM_WRITE_PERMISSION_0		0x84058
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_0			0x84450
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_0			0x84458
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_1			0x84470
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_1			0x84478
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_2			0x84490
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_2			0x84498
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_3			0x844B0
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_3			0x844B8
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_4			0x844D0
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_4			0x844D8
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_5			0x844F0
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_5			0x844F8
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_6			0x84510
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_6			0x84518
+#define REG0_TDDR_PM_TDDR_PM_READ_PERMISSION_7			0x84530
+#define REG0_TDDR_PM_TDDR_PM_WRITE_PERMISSION_7			0x84538
+#define REG0_TPCTL_PM_TPCTL_PM_READ_PERMISSION_0		0x84850
+#define REG0_TPCTL_PM_TPCTL_PM_WRITE_PERMISSION_0		0x84858
+#define REG0_TPUB_PM_TPUB_PM_READ_PERMISSION_0			0x84C50
+#define REG0_TPUB_PM_TPUB_PM_WRITE_PERMISSION_0			0x84C58
+#define REG0_TPON_PM_TPON_PM_READ_PERMISSION_0			0x85050
+#define REG0_TPON_PM_TPON_PM_WRITE_PERMISSION_0			0x85058
+#define REG0_TGSWIP_PM_TGSWIP_PM_READ_PERMISSION_0		0x85450
+#define REG0_TGSWIP_PM_TGSWIP_PM_WRITE_PERMISSION_0		0x85458
+#define REG0_TQSPIC_PM_TQSPIC_PM_READ_PERMISSION_0		0x85850
+#define REG0_TQSPIC_PM_TQSPIC_PM_WRITE_PERMISSION_0		0x85858
+#define REG0_TMSI_PM_TMSI1_PM_READ_PERMISSION_0			0x85C50
+#define REG0_TMSI_PM_TMSI1_PM_WRITE_PERMISSION_0		0x85C58
+#define REG0_TDMAT2_PM_TDMAT2_PM_READ_PERMISSION_0		0x86050
+#define REG0_TDMAT2_PM_TDMAT2_PM_WRITE_PERMISSION_0		0x86058
+#define REG0_TQSPID_PM_TQSPID_PM_READ_PERMISSION_0		0x86450
+#define REG0_TQSPID_PM_TQSPID_PM_WRITE_PERMISSION_0		0x86458
+#define REG0_TDMAR1_PM_TDMAR1_PM_READ_PERMISSION_0		0x86850
+#define REG0_TDMAR1_PM_TDMAR1_PM_WRITE_PERMISSION_0		0x86858
+#define REG0_TDMAR2_PM_TDMAR2_PM_READ_PERMISSION_0		0x86C50
+#define REG0_TDMAR2_PM_TDMAR2_PM_WRITE_PERMISSION_0		0x86C58
+#define REG0_TACA_PM_TACA_PM_READ_PERMISSION_0			0x87050
+#define REG0_TACA_PM_TACA_PM_WRITE_PERMISSION_0			0x87058
+#define REG0_TPCIE_CR1_PM_TPCIE_CR1_PM_READ_PERMISSION_0	0x87450
+#define REG0_TPCIE_CR1_PM_TPCIE_CR1_PM_WRITE_PERMISSION_0	0x87458
+#define REG0_TPCI2_PM_TPCI2_PM_READ_PERMISSION_0		0x87850
+#define REG0_TPCI2_PM_TPCI2_PM_WRITE_PERMISSION_0		0x87858
+#define REG0_TPCI1_PM_TPCI1_PM_READ_PERMISSION_0		0x87C50
+#define REG0_TPCI1_PM_TPCI1_PM_WRITE_PERMISSION_0		0x87C58
+#define REG0_TMSI2_PM_TMSI2_PM_READ_PERMISSION_0		0x88050
+#define REG0_TMSI2_PM_TMSI2_PM_WRITE_PERMISSION_0		0x88058
+#define REG0_TPCIE_CR2_PM_TPCIE_CR2_PM_READ_PERMISSION_0	0x88850
+#define REG0_TPCIE_CR2_PM_TPCIE_CR2_PM_WRITE_PERMISSION_0	0x88858
+#define REG0_TLNPPV4_PM_TLNPPV4_PM_READ_PERMISSION_0		0x92050
+#define REG0_TLNPPV4_PM_TLNPPV4_PM_WRITE_PERMISSION_0		0x92058
+#define REG0_TXPCS_PM_TXPCS_PM_READ_PERMISSION_0		0x92450
+#define REG0_TXPCS_PM_TXPCS_PM_WRITE_PERMISSION_0		0x92458
+#define REG0_TPC1DBI_PM_TPCI1DBI_PM_READ_PERMISSION_0		0x92850
+#define REG0_TPC1DBI_PM_TPCI1DBI_PM_WRITE_PERMISSION_0		0x92858
+#define REG0_TPC2DBI_PM_TPCI2DBI_PM_READ_PERMISSION_0		0x92C50
+#define REG0_TPC2DBI_PM_TPCI2DBI_PM_WRITE_PERMISSION_0		0x92C58
+#define REG0_TSSB_PM_TSSB_PM_READ_PERMISSION_0			0x93050
+#define REG0_TSSB_PM_TSSB_PM_WRITE_PERMISSION_0			0x93058
+#define REG0_TROM_PM_TROM_PM_READ_PERMISSION_0			0x93450
+#define REG0_TROM_PM_TROM_PM_WRITE_PERMISSION_0			0x93458
+#define REG0_TMACS_PM_TMAC_PM_READ_PERMISSION_0			0x95050
+#define REG0_TMACS_PM_TMAC_PM_WRITE_PERMISSION_0		0x95058
+#define REG0_TIOCU1_PM_TIOCU1_PM_READ_PERMISSION_0		0x95450
+#define REG0_TIOCU1_PM_TIOCU1_PM_WRITE_PERMISSION_0		0x95458
+#define REG0_TDM4_PM_TDM4_PM_READ_PERMISSION_0			0x95850
+#define REG0_TDM4_PM_TDM4_PM_WRITE_PERMISSION_0			0x95858
+
+#endif
diff --git a/arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx4.h b/arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx4.h
new file mode 100644
index 000000000000..a2fbe8ce5221
--- /dev/null
+++ b/arch/mips/lantiq/grx500_bootcore/ngi_prx300_ssx4.h
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ */
+
+#ifndef _SSX4_SSX_H
+#define _SSX4_SSX_H
+
+/*! Base Address of SSX4_SHARED_LINK */
+#define SSX4_SHARED_LINK_MODULE_BASE 0x16000000u
+
+/*! Permission Offset (relative) */
+#define REG4_TREG4_PM_TREG4_PM_READ_PERMISSION_0 0x80050
+#define REG4_TREG4_PM_TREG4_PM_WRITE_PERMISSION_0 0x80058
+#define REG4_TREG4_PM_TREG4_PM_READ_PERMISSION_1 0x80070
+#define REG4_TREG4_PM_TREG4_PM_WRITE_PERMISSION_1 0x80078
+#define REG4_TCGU_PM_TCGU_PM_READ_PERMISSION_0 0x80450
+#define REG4_TCGU_PM_TCGU_PM_WRITE_PERMISSION_0 0x80458
+#define REG4_TRCU_PM_TRCU_PM_READ_PERMISSION_0 0x80850
+#define REG4_TRCU_PM_TRCU_PM_WRITE_PERMISSION_0 0x80858
+#define REG4_TPMU_PM_TPMU_PM_READ_PERMISSION_0 0x80C50
+#define REG4_TPMU_PM_TPMU_PM_WRITE_PERMISSION_0 0x80C58
+#define REG4_TPCM_PM_TPCM_PM_READ_PERMISSION_0 0x81050
+#define REG4_TPCM_PM_TPCM_PM_WRITE_PERMISSION_0 0x81058
+#define REG4_TGPIO_PM_TGPIO_PM_READ_PERMISSION_0 0x81450
+#define REG4_TGPIO_PM_TGPIO_PM_WRITE_PERMISSION_0 0x81458
+#define REG4_TLEDC_PM_TLEDC_PM_READ_PERMISSION_0 0x81850
+#define REG4_TLEDC_PM_TLEDC_PM_WRITE_PERMISSION_0 0x81858
+#define REG4_TI2C_PM_TI2C0_PM_READ_PERMISSION_0 0x81C50
+#define REG4_TI2C_PM_TI2C0_PM_WRITE_PERMISSION_0 0x81C58
+#define REG4_TASC0_PM_TASC0_PM_READ_PERMISSION_0 0x82050
+#define REG4_TASC0_PM_TASC0_PM_WRITE_PERMISSION_0 0x82058
+#define REG4_TASC1_PM_TASC1_PM_READ_PERMISSION_0 0x82450
+#define REG4_TASC1_PM_TASC1_PM_WRITE_PERMISSION_0 0x82458
+#define REG4_TGPT0_PM_TGPT0_PM_READ_PERMISSION_0 0x82850
+#define REG4_TGPT0_PM_TGPT0_PM_WRITE_PERMISSION_0 0x82858
+#define REG4_TSSC0_PM_TSSC0_PM_READ_PERMISSION_0 0x82C50
+#define REG4_TSSC0_PM_TSSC0_PM_WRITE_PERMISSION_0 0x82C58
+#define REG4_TSSC1_PM_TSSC1_PM_READ_PERMISSION_0 0x83050
+#define REG4_TSSC1_PM_TSSC1_PM_WRITE_PERMISSION_0 0x83058
+#define REG4_TDMA0_PM_TDM0_PM_READ_PERMISSION_0 0x83450
+#define REG4_TDMA0_PM_TDM0_PM_WRITE_PERMISSION_0 0x83458
+#define REG4_TGPT1_PM_TGPT1_PM_READ_PERMISSION_0 0x83C50
+#define REG4_TGPT1_PM_TGPT1_PM_WRITE_PERMISSION_0 0x83C58
+#define REG4_TGPT2_PM_TGPT2_PM_READ_PERMISSION_0 0x84050
+#define REG4_TGPT2_PM_TGPT2_PM_WRITE_PERMISSION_0 0x84058
+#define REG4_TI2C0_PM_TI2C1_PM_READ_PERMISSION_0 0x85050
+#define REG4_TI2C0_PM_TI2C1_PM_WRITE_PERMISSION_0 0x85058
+#define REG4_TI2C1_PM_TI2C2_PM_READ_PERMISSION_0 0x85450
+#define REG4_TI2C1_PM_TI2C2_PM_WRITE_PERMISSION_0 0x85458
+
+
+#endif
